packages/flutter/lib/src/widgets/binding.dart

```diff
@@ -943,10 +943,50 @@ class RenderObjectToWidgetElement<T extends RenderObject> extends RootRenderObje
   }
 }
 
+/// Base class that provide necessary information to register a ServiceExtension.
+abstract class ExtensionRegistrarBase {
+  /// Name to registers a service extension method with.(full name
+  /// "ext.flutter.name").
+  String get name;
+
+  /// Callback called when the extension method is called.
+  ServiceExtensionCallback get callback;
+}
+
+/// Binding for service extensions to register.
+///
+/// There might be multiple optional ServiceExtensions in practice.
+/// For example, ServiceExtension used by 'flutter driver' is valid only in that command.
+/// Using [ExtensionBinding.registerExtensionRegistrar], developers can register custom service extensions conditionally,
+/// without inheriting & instancing a [BindingBase] instance.
+mixin ExtensionBinding on BindingBase {
+  /// List to save ExtensionRegistrars.
+  static final List<ExtensionRegistrarBase> _extensionRegistrars = <ExtensionRegistrarBase>[];
```

Comment:
Seems like we could make this simpler (removing the ExtensionRegistrarBase class entirely) if we make this a map of name to callback, instead of a list of objects.

Reply:
Though a simple kv could make things simpler, I would prefer to keep this design. As we may do some extra logic in a ExtensionRegistrar(like _DriverExtensionRegistrar), not only a name and callback.